### ISSUE TITLE
strategy for duplicate custom element definitions

### ISSUE THREAD

**Issue by geoffrich** [AlaskaAirlines/WC-Generator#159]
## Is your feature request related to a problem? Please describe.

With the growth of the design system, we are starting to run into issues where multiple versions of the same element are present on the page, e.g. https://github.com/AlaskaAirlines/auro-button/issues/119#issuecomment-867902380. This causes issues, since you can only define a custom element once. Whoever defines the custom element first wins, and all instances of that element will use the first definition instead of what might be imported by another script. This will cause issues when one script is importing a version of the element with a bug or when different major versions of an element are loaded.

At the moment this is a silent issue, since we make sure the custom element is not defined before defining it. This prevents errors from duplicate define calls; however, it also makes these issues harder to identify because it silently uses the other version.

This is primarily an issue with consuming apps, i.e. there are multiple front-end components present on a page with different versions of Auro elements. However, it could also become an issue for Auro components that have nested custom element dependencies, e.g. auro-loader inside of auro-button.

## Describe the solution you'd like
We may be able to adopt or encourage the adoption of the [Scoped Custom Element Registries polyfill](https://open-wc.org/docs/development/scoped-elements/#motivation). I believe this would only apply inside another shadow root.

Another workaround could be exporting the custom element definition so consumers could define their own tag name if they run into this issue, e.g. `auro-button-v1` and `auro-button-v2`.

## Describe alternatives you've considered
We could do nothing and document the issue. Teams would need to coordinate if they both plan on using Auro components and synchronize updates.

## Additional context
Here's some of Adobe's thoughts on the issue: https://github.com/adobe/spectrum-web-components/issues/1024

We should also research how/if other WC libraries handle this problem.

**Comment by blackfalcon** [comment]
@geoffrich and I discussed this issue in length and came to the following conclusion. 

Defining a custom element can happen only once within a view. Because of this, we need to be careful about the loading order of CEs within a view. We seem to be running into the primary issue with Component Hanger CEs, mainly because consumers have no control over the versions of CEs nested within. E.g. a consumer can update their local packages, like the version of auro-button, but not the version of auro-button nested inside the Component Hanger CE.

@geoffrich was able to [troubleshoot](https://github.com/AlaskaAirlines/auro-button/issues/119#issuecomment-867843836) this with @liambrozik and what worked was to delay the loading of the Component Hanger component until AFTER all the local definitions of Auro CEs were loaded. Given this scenario, the local version wins out and the nested version is ignored. 

This is not a Component Hanger issue alone. There is the potential for this issue to be seen in any Auro CE that has a nested dependency. E.g. auro-button that has a dependency on auro-loader. 

**Potential issue:** Given this scenario, this places greater importance on the Auro team to be diligent in its releasing of breaking changes. E.g. an app that has a dependency on a new major version of a CE with a breaking API change and using another CE or CH CE that has a dependency on a lesser version. Given the loading strategy above, loading the major version would break the nested lesser version. 

Potential solution(s): One thing Auro can consider is to never nest other CEs as dependencies. E.g. auro-button would be updated to include auro-loader as a peer dependency. In most cases, this will not be an issue and the primary update is to include the additional installer notes for consumers. Where this may be unwelcome is with Auro's CDN solution. Without the nested CE as a dependency, then a consumer would be required to import additional bundled CEs. 

In regard to Component Hanger, the same idea could be discussed where the Auro CE is not included with its deliverable and the consumers would be required to either import from Auro's CDN or import into the app locally as a peer dependency. 

This breaks the model of "install once and done", but _rock meets hardplace_. 

I open this to @gusnaughton @geoffrich @jason-capsule42 for further comments.

**Comment by blackfalcon** [comment]
Added note: @geoffrich proposal of using the registry is being put on the back-burner as this is highly experimental and an effort we currently cannot take on. The decision was to keep an eye on that spec over time and adapt when appropriate. 

His other idea of exporting the JS class of a component was also discussed and there is a lot of effort there and we are unclear as to if this will actually be used. This scenario is not applicable until we are forced into this as a solution.

**Comment by blackfalcon** [comment]
@gusnaughton @jason-capsule42 please see https://github.com/AlaskaAirlines/WC-Generator/issues/159#issuecomment-868752049

This has been open for a while, looking for feedback.

**Comment by blackfalcon** [comment]
Hearing no feedback from other peers, I feel we are safe to move forward on this and codify a document.

**Comment by blackfalcon** [comment]
This issue has been open for a long time in order to really review its impact on building web components. 

Having encountered another issue where a nested version of auro-header in auro-card was keeping the proper version of things rendering in the Auro doc site has led me to the conclusion that we need this as a policy. 

> Potential solution(s): One thing Auro can consider is to never nest other CEs as dependencies. E.g. auro-button would be updated to include auro-loader as a peer dependency. In most cases, this will not be an issue and the primary update is to include the additional installer notes for consumers. Where this may be unwelcome is with Auro's CDN solution. Without the nested CE as a dependency, then a consumer would be required to import additional bundled CEs.

**Comment by blackfalcon** [comment]
In addition to this policy, we need to ensure that primitive elements are never added as nested dependencies with any given element. They should always be peer dependencies allowing the app devs to have control over all the versions of elements in their app.

**Comment by geoffrich** [comment]
Even if they are peer deps, I think the bundled versions (which the doc site uses) will still bundle all dependencies.  We'll also need to make changes to the bundle config to _not_ bundle other custom elements.

(Though if we do that, it's a breaking change for the bundle setup -- consumers of the bundles will need to know what other bundled custom elements they need to load. e.g., if you load auro-flight__bundled, you also need to load auro-flightline__bundled. This is different to how the bundle works today, and might be confusing? Though it would prevent the nested custom element dep problem for bundles.)

**Comment by blackfalcon** [comment]
@geoffrich yeah, we would need to look into that. I would propose that maybe we come up with a deprecation strategy where we maintain a fully bundled version and a separate bundle version - that way we can avoid a breaking change release.

**Comment by blackfalcon** [comment]
From @gusnaughton in a text

> What if we detected version collisions before the customElement.register() call that checks a globally maintained map of auro component names to unique versions needed by the Dom with a dynamically generated namespace that gets prepended to the custom element name and adjust the consuming code to the new element name?
